eventuate-core/src/main/scala/com/rbmhtechnology/eventuate/Recovery.scala

```diff
@@ -81,57 +83,102 @@ private class Recovery(endpoint: ReplicationEndpoint) {
+
+  private def readLogSequenceNrs: Future[Map[String, Long]] =
+    readEventLogClocks.map(_.mapValues(_.sequenceNr).view.force)
 
   /**
    * Reads the clocks from local event logs.
    */
-  def readEventLogClocks: Future[Map[String, EventLogClock]] = {
-    println(s"[recovery of ${endpoint.id}] Read clocks from local event logs ...")
-    readClocksImpl
+  def readEventLogClocks: Future[Map[String, EventLogClock]] =
+    Future.traverse(endpoint.logNames)(name => readEventLogClock(endpoint.logs(name)).map(name -> _)).map(_.toMap)
+
+  /**
+   * Synchronize sequence numbers of local logs with replication progress stored in remote replicas.
+   * @return a set of [[ReplicationEndpointInfo]] of all remote replicas
+   */
+  def synchronizeReplicationProgressesWithRemote(info: ReplicationEndpointInfo): Future[Set[ReplicationEndpointInfo]] =
+    Future.sequence(endpoint.connectors.map(connector => synchronizeReplicationProgressWithRemote(connector.remoteAcceptor, info)))
+
+  private def synchronizeReplicationProgressWithRemote(remoteAcceptor: ActorSelection, info: ReplicationEndpointInfo): Future[ReplicationEndpointInfo] =
+    readResult[SynchronizeReplicationProgressSuccess, SynchronizeReplicationProgressFailure, ReplicationEndpointInfo](
+      Retry(remoteAcceptor.ask(SynchronizeReplicationProgress(info)), remoteOperationRetryDelay, remoteOperationRetryMax), _.info, _.cause)
+
+  /**
+   * Update the locally stored replication progress of remote replicas with the sequence numbers given in ``info``.
+   * Replication progress that is greater than the corresponding sequence number in ``info`` is reset to that
+   */
+  def synchronizeReplicationProgress(info: ReplicationEndpointInfo): Future[Unit] = {
+    Future.traverse(endpoint.commonLogNames(info)) { name =>
+      val logActor = endpoint.logs(name)
+      val logId = info.logId(name)
+      val remoteSequenceNr = info.logSequenceNrs(name)
+      for {
+        currentProgress <- readReplicationProgress(logActor, logId)
+        _ <- if (currentProgress > remoteSequenceNr) updateReplicationProgress(logActor, logId, info.logSequenceNrs(name))
+        else Future.successful(currentProgress)
+      } yield ()
+    } map (_ => ())
   }
 
-  private def readClocksImpl: Future[Map[String, EventLogClock]] = {
-    Future.traverse(endpoint.logNames)(name => readEventLogClock(endpoint.logs(name)).map(name -> _)).map(_.toMap)
+  private def readReplicationProgress(logActor: ActorRef, logId: String): Future[Long] =
+    readResult[GetReplicationProgressSuccess, GetReplicationProgressFailure, Long](
+      logActor.ask(GetReplicationProgress(logId)), _.storedReplicationProgress, _.cause)
+
+  private def updateReplicationProgress(logActor: ActorRef, logId: String, sequenceNr: Long): Future[Long] = {
+    readResult[ReplicationWriteSuccess, ReplicationWriteFailure, Long](
+      logActor.ask(ReplicationWrite(Seq.empty, sequenceNr, logId, VectorTime.Zero)), _.storedReplicationProgress, _.cause)
```

Comment:
I see, ok. But setting it to `VectorTime.Zero` makes only sense if the remote location lost all events. If it has been restored from a backup, we would potentially transfer a large number of events just for being ignored at the target.